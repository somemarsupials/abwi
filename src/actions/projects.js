import { api } from '../App';

export const actions = {
  FETCHING: 'PROJECT_FETCHING',
  FETCH_SUCCESS: 'PROJECT_FETCH_SUCCESS',
  FETCH_FAIL: 'PROJECT_FETCH_FAIL',
};

async function makeRequest(id) {
  return await fetch(`${api}/projects/${id}?detail=true`);
};

export function projectLoading(bool) {
  return {
    type: actions.FETCHING,
    isFetching: bool,
  };
};

export function projectFetched(data) {
  return {
    type: actions.FETCH_SUCCESS,
    data: data,
  };
};

export function projectNotFetched(error) {
  return {
    type: actions.FETCH_FAIL,
    error: error,
  };
};

export function fetchProject(id) {
  return async (dispatch) => {
    let response, data, error;
    dispatch(projectLoading(true));

    try {
      response = await makeRequest(id);
    } 
    catch (e) {
      error = e.message;
    };

    dispatch(projectLoading(false));
    error = error || (!response.ok && response.status);

    if (!error) {
      data = await response.json();
    };

    if (error) {
      dispatch(projectNotFetched(error));
    } else {
      dispatch(projectFetched(data));
    };
  };
};
