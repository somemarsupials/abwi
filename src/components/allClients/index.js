export { default as ClientsPage } from './clientsList';
export { default as ClientsList } from './clientsPage';
