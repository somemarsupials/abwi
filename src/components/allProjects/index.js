export { default as List } from './list';
export { default as Page } from './page';
