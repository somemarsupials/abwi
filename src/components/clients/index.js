export { default as ClientPage } from './clientPage';
export { default as ClientDetail } from './clientDetail';
