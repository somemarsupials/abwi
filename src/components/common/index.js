export { default as TitleBar } from './titleBar';
export { default as ErrorBar } from './errorBar';
export { default as InfoBar } from './infoBar';
export { default as SectionBar } from './sectionBar';
