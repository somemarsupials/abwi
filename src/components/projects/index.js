export { default as ProjectPage } from './projectPage';
export { default as ProjectDetail } from './projectDetail';
