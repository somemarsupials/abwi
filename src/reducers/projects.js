import { combineReducers } from 'redux';
import { actions } from '../actions/projects';
import { actions as modalActions } from '../actions/projectModal';

function fetching(state = false, action) {
  switch (action.type) {
    case actions.FETCH_LOADING:
      return action.active;
    default:
      return state;
  }
};

const initialState = {
  error: null,
  data: [],
};

function fetched(state = initialState, action) {
  switch (action.type) {
    case actions.FETCHED:
      return {
        error: action.error,
        data: action.data,
      };
    case modalActions.CREATED:
      return {
        error: state.error,
        data: state.data.concat([action.data]),
      };
    default:
      return state;
  }
};

export default combineReducers({
  fetching: fetching,
  response: fetched,
});
