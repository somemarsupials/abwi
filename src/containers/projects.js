import React, { Component } from 'react';
import { connect } from 'react-redux';

import { projectActions } from '../actions';
import { ProjectsOverview } from '../components/projects';


export class Projects extends Component {
  async componentDidMount() {
    let { api, success, fail } = this.props;
    this.props.fetch(api, success, fail);
  };

  render() {
    return (
      <ProjectsOverview
        projects={this.props.projects} 
        error={this.props.error}
      />
    );
  };
};


function mapStateToProps(state) {
  return { 
    projects: state.projects.data,
    error: state.projects.error,
  };
};


function mapDispatchToProps(dispatch) {
  let { fetchedProjects, projectFetchFailed } = projectActions;
  return {
    success: (projects) => {
      dispatch(fetchedProjects(projects))
    },
    fail: (error) => {
      dispatch(projectFetchFailed(error))
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Projects);
