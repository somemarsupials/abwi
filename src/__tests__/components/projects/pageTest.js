import React from 'react';
import { shallow } from 'enzyme';

import { Page } from '../../../components/projects';

describe('Page', () => {
  let page;
  let project;

  beforeEach(() => {
    project = 'project';
  });

  describe('with error', () => {
    beforeEach(() => {
      page = shallow(<Page error={'an error'}/>
      );
    });

    it('renders with error bar', () => {
      expect(page).toMatchSnapshot();
    });
  });

  describe('when loading', () => {
    beforeEach(() => {
      page = shallow(<Page project={null} fetching={true} error={null} />);
    });

    it('renders with loading information', () => {
      expect(page).toMatchSnapshot();
    });
  });

  describe('when finished loading', () => {
    beforeEach(() => {
      page = shallow(<Page loading={false} project={project} />
      );
    });

    it('correctly renders', () => {
      expect(page).toMatchSnapshot();
    });
  });
});

