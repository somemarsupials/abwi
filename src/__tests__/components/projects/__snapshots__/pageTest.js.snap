// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page when finished loading correctly renders 1`] = `
<div>
  <Component
    text={null}
  />
  <Component
    project="project"
  />
</div>
`;

exports[`Page when loading renders with loading information 1`] = `
<div>
  <Component
    text={null}
  />
  <Component
    text="Fetching project..."
  />
  <Component
    detail={null}
    text="Could not load project"
  />
</div>
`;

exports[`Page with error renders with error bar 1`] = `
<div>
  <Component />
  <Component
    detail="an error"
    text="Could not load project"
  />
  <Component />
</div>
`;
